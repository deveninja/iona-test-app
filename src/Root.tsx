import { Provider } from 'react-redux';
import Store from 'store';

/**
 * @description - This will be used to improve testing suites and
 * as single source of provider wrapper
 * @param param0 - children props
 * @returns Provider wrapper for all children
 */
const Root: React.FC = ({ children }) => {
    return (
        <Provider store={Store}>
            {children}
        </Provider>
    )
}

export default Root