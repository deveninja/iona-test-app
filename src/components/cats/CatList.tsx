import { getCatBreed, getCatBreedList } from 'actions'
import CardComponent from 'components/common/card/Card'
import DropDown from 'components/common/dropdown/DropDownList'
import SpinnerComponent from 'components/common/snipper/Spinner'
import { Breed } from 'enums/breed'
import { AppRoute } from 'enums/route'
import { useMemo, useEffect, useState, useCallback } from 'react'
import { Alert, Button, Col, Row } from 'react-bootstrap'
import { useSelector, useDispatch } from 'react-redux'
import { RouteComponentProps } from 'react-router-dom'
import { AppState } from 'store'

import styles from './cat.module.scss'

const CatList: React.FC<RouteComponentProps> = (props: RouteComponentProps) => {
    const [currentBreed, setCurrentBreed] = useState('')
    const [limit,] = useState(10)
    const catsBreedList: any = useSelector<AppState, any>(state => state.breedsReducer[Breed.CATS])
    const cats = useSelector<AppState, any>(state => state.catsReducer)
    const loading = useSelector<AppState, boolean>(state => state.globalReducer.loadingState)

    const dispatch = useDispatch()

    const {
        history,
        match: {
            params: {
                breed
            }
        }
    }: any = props


    // Memoise 
    const showBtn = useMemo(() => {
        const len = cats[currentBreed] ? Object.values(cats[currentBreed]).length : 0
        return catsBreedList?.[breed]?.page > 3 || len >= 15 || catsBreedList?.[breed]?.final

    }, [catsBreedList, breed, cats, currentBreed])


    const page = useMemo(() => (catsBreedList?.[breed]?.page ?? 1), [catsBreedList, breed])
    const options = useMemo(() => (catsBreedList ?? []), [catsBreedList])


    // Renders the cat card components
    const renderCats = useCallback((data: any) => {
        return (
            Object.values(data).map((i: any) => {
                return (
                    <CardComponent
                        key={i?.id}
                        imgSrc={i?.url}
                        size="23%"
                        btnLabel={`Have a closer look`}
                        link={{
                            to: `${AppRoute.Cats}/${currentBreed}/${i?.id}`,
                            prevPath: `${AppRoute.Cats}/${currentBreed}`
                        }}
                    />
                )
            })
        )
    }, [currentBreed])


    // onchange handler that sets the current breed selected and modifies
    // the route
    const onSelectBreed = (e: Event) => {
        const { value }: any = e.target
        setCurrentBreed(value)
        history.push(`${AppRoute.Cats}/${value}`)
    }


    useEffect(() => {

        if (!Object.values(catsBreedList).length) {
            dispatch(getCatBreedList())
        }

    }, [dispatch, catsBreedList])


    /**
     * =======================================================================
     * @description - This block will solve if the breed view should call
     * the api endpoint or not without adding complexity in the action types
     * inside the reducer
     * =======================================================================
     */
    useEffect(() => {
        if (currentBreed) {
            if (!cats?.[currentBreed]) {
                dispatch(getCatBreed({ id: currentBreed, limit, page }))
            }

        }

    }, [dispatch, currentBreed, cats, limit, page])
    /** ===================================================================== */


    useEffect(() => {
        setCurrentBreed(breed)
    }, [breed])



    /**
     * @SPECS
     * data-testid="pageHeading"
     * data-testid="breedDropdown"
     * data-testid="breedMoreBtn"
     * data-testid="returnHomeBtn"
     * data-testid="listSection"
     */

    return (
        <div className={styles.catList}>
            <Row>
                <span data-testid="pageHeading" className={styles.title}>Cat Browser</span>
            </Row>
            <div className={styles.wrapper}>
                <Row>
                    <Col lg={3}>
                        <div className={`${styles.paddedFlex} ${styles.selectArea}`}>
                            <div>
                                <DropDown
                                    data-testid="breedDropdown"
                                    options={Object.values(options)}
                                    onChange={onSelectBreed}
                                    selected={currentBreed}
                                />
                                {
                                    !showBtn &&
                                    breed &&
                                    <Button
                                        data-testid="breedMoreBtn"
                                        variant="secondary"
                                        onClick={() => dispatch(
                                            getCatBreed({
                                                id: currentBreed,
                                                page,
                                            })
                                        )}
                                    >
                                        Load more {`${catsBreedList[breed]?.name} breed` ?? 'cats'}
                                    </Button>
                                }

                            </div>
                            <Button
                                data-testid="returnHomeBtn"
                                variant="secondary"
                                onClick={() => history.push(`${AppRoute.Home}`)}
                            >
                                Back to Home
                            </Button>
                        </div>
                    </Col>
                    <Col lg={9}>
                        <div
                            data-testid="listSection"
                            className={`${styles.paddedFlex} 
                            ${styles.contentArea}`}
                        >
                            {
                                // If the result has length of more then 0
                                cats?.[currentBreed] &&
                                Object.values(cats?.[currentBreed]).length > 0 &&
                                renderCats(cats[currentBreed])
                            }

                            {
                                // If After calling the api and the the result count
                                // is ZERO
                                cats?.[currentBreed] &&
                                Object.values(cats?.[currentBreed]).length < 1 &&
                                <Alert key={Date.now()} variant={'warning'}>
                                    No Cats found in our database.
                                </Alert>
                            }

                            {
                                // While calling the endpoint
                                loading && !cats?.[currentBreed]?.length &&
                                <SpinnerComponent message="Loading list" />
                            }

                        </div>
                    </Col>
                </Row>
            </div>
        </div>
    )
}

export default CatList
