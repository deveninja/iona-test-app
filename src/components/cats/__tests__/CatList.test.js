import renderer from 'react-test-renderer'
import Root from 'Root'
import CatList from '../CatList'
import { mount } from 'enzyme';
import { render, cleanup } from '@testing-library/react';


describe('<CatList />', () => {

    const match = { params: { breed: '' }, isExact: true, path: "", url: "" }
    const Element = <Root><CatList match={match} /></Root>
    let component
    let componentSnaphot


    beforeEach(() => {
        component = render(Element)
        componentSnaphot = mount(Element)
    })

    afterEach(() => {
        componentSnaphot.unmount()
        cleanup()
    })

    it('displays the pageHeading section', () => {
        const section = component.getByTestId('pageHeading')
        expect(section).toBeTruthy()
    })


    it('displays the breedDropdown section', () => {
        const section = component.getByTestId('breedDropdown')
        expect(section).toBeTruthy()
    })

    it('displays the returnHomeBtn section', () => {
        const section = component.getByTestId('returnHomeBtn')
        expect(section).toBeTruthy()
    })

    it('displays the listSection section', () => {
        const section = component.getByTestId('listSection')
        expect(section).toBeTruthy()
    })

    it('matches snapshot', () => {
        const element = renderer.create(Element).toJSON()
        expect(element).toMatchSnapshot()
    })
})
