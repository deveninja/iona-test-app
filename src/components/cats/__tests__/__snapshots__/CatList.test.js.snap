// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatList /> matches snapshot 1`] = `
<div
  className="catList"
>
  <div
    className="row"
  >
    <span
      className="title"
      data-testid="pageHeading"
    >
      Cat Browser
    </span>
  </div>
  <div
    className="wrapper"
  >
    <div
      className="row"
    >
      <div
        className="col-lg-3"
      >
        <div
          className="paddedFlex selectArea"
        >
          <div>
            <select
              aria-label="Default select example"
              className="dropDown form-select"
              data-testid="breedDropdown"
              onChange={[Function]}
              value=""
            >
              <option
                disabled={true}
                value=""
              >
                Select Breed
              </option>
            </select>
            
          </div>
          <button
            className="btn btn-secondary"
            data-testid="returnHomeBtn"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Back to Home
          </button>
        </div>
      </div>
      <div
        className="col-lg-9"
      >
        <div
          className="paddedFlex 
                            contentArea"
          data-testid="listSection"
        >
          <div
            className="spinnerComponentWrapper"
          >
            <div
              className="spinnerWrapper"
            >
              <div
                className="spinner-grow text-secondary"
              />
              <div
                className="spinner-grow text-secondary"
              />
              <div
                className="spinner-grow text-secondary"
              />
              <div
                className="spinner-grow text-secondary"
              />
              <div
                className="spinner-grow text-secondary"
              />
              <div
                className="spinner-grow"
                role="status"
              >
                <span
                  className="visually-hidden"
                >
                  Loading...
                </span>
              </div>
            </div>
            <h1>
              Loading Cat Breeds...
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
